apollo-router/src/services/layers/persisted_queries/manifest_poller.rs

```diff
@@ -0,0 +1,364 @@
+    ready_sender: mpsc::Sender<ManifestPollResultOnStartup>,
+    mut shutdown_event_receiver: mpsc::Receiver<()>,
+    http_client: Client,
+) {
+    let mut uplink_executor = stream::select_all(vec![
+        stream_from_uplink::<PersistedQueriesManifestQuery, MaybePersistedQueriesManifestChunks>(
+            uplink_config.clone(),
+        )
+        .filter_map(|res| {
+            let http_client = http_client.clone();
+            let graph_ref = uplink_config.apollo_graph_ref.clone();
+            async move {
+                match res {
+                    Ok(Some(chunks)) => match manifest_from_chunks(chunks, http_client).await {
+                        Ok(new_manifest) => Some(ManifestPollEvent::NewManifest(new_manifest)),
+                        Err(e) => Some(ManifestPollEvent::FetchError(e)),
+                    },
+                    Ok(None) => Some(ManifestPollEvent::NoPersistedQueryList { graph_ref }),
+                    Err(e) => Some(ManifestPollEvent::Err(e.into())),
+                }
+            }
+        })
+        .boxed(),
+        shutdown_event_receiver
+            .recv()
+            .into_stream()
+            .filter_map(|res| {
+                future::ready(match res {
+                    Some(()) => Some(ManifestPollEvent::Shutdown),
+                    None => Some(ManifestPollEvent::Err(
+                        anyhow!("could not receive shutdown event for persisted query layer")
+                            .into(),
+                    )),
+                })
+            })
+            .boxed(),
+    ])
+    .take_while(|msg| future::ready(!matches!(msg, ManifestPollEvent::Shutdown)))
+    .boxed();
+
+    let mut resolved_first_pq_manifest = false;
+
+    while let Some(event) = uplink_executor.next().await {
+        match event {
+            ManifestPollEvent::NewManifest(new_manifest) => {
+                // copy all of the bodies in the new manifest into a hash set before acquiring any locks
+                let new_bodies: HashSet<String> = new_manifest.values().cloned().collect();
+                existing_persisted_query_manifest
+                    .write()
+                    .map(|mut existing_manifest| {
+                        existing_persisted_query_bodies
+                            .write()
+                            .map(|mut existing_bodies| {
+                                // update the existing map of pq id to pq body to be the new collection
+                                *existing_manifest = new_manifest;
+
+                                // update the set of pq bodies from the values in the new collection
+                                *existing_bodies = new_bodies;
+                            })
+                            .unwrap_or_else(|e| {
+                                panic!(
+                                    "could not acquire write lock on persisted query body set: {e}"
+                                )
```

Comment:
it's in a separate thread, so there might be a way to recover from that 🤔

Reply:
i'd love to see how this would work! no clue how i'd do that 😆

Reply:
Usually for locks we use `expect` as recovery from poisoning is is very difficult to do safely and introduces unexpected lifecycle handling. i.e. you have to throw everything away. Honestly I think that we should never try to recover from poisoned locks.

Reply:
all right, wfm!